File: 2516_take_k_of_each_character_from_left_and_right.cpp
class Solution {
public:
    int takeCharacters(string s, int k) {
        int n = s.size(), freq[3] = {}, ans = n, left = 0;
        for (auto c : s) freq[c - 'a']++;
        if (*min_element(freq, freq + 3) < k) return -1;
        for (int right = 0; right < n; right++) {
            freq[s[right] - 'a']--;
            while (left <= right && *min_element(freq, freq + 3) < k) freq[s[left++] - 'a']++;
            ans = min(ans, n - (right - left + 1));
        }
        return ans;
    }
};

File: 2516_take_k_of_each_character_from_left_and_right.java
class Solution {
    public int takeCharacters(String s, int k) {
        int n = s.length(), l = 0, r = 0, minLen = n;
        int[] counts = new int[3];
        for (char c : s.toCharArray()) counts[c - 'a']++;
        if (counts[0] < k || counts[1] < k || counts[2] < k) return -1;
        for (; r < n; r++) {
            counts[s.charAt(r) - 'a']--;
            while (counts[0] < k || counts[1] < k || counts[2] < k)
                counts[s.charAt(l++) - 'a']++;
            minLen = Math.min(minLen, n - (r - l + 1));
        }
        return minLen;
    }
}
